### ISSUE TITLE
Out of memory with tuner.image_to_text

### ISSUE THREAD

**Issue by gr0esch** [MaartenGr/BERTopic#1464]
Hey, first of all – many tanks for the great package – especially with the update for multimodality, which I am using it for here.

I have image only dataset that I try to run a topic model on. I pre-calculated the embeddings on the GPU and am now trying to build my complete model. Dimensionality reduction and clustering seems to work, but I do get a out of memory at the representation stage while the algorithm is trying to convert the images to text. Is there any way around that?

Currently, I am running the code on a GPU (T4) with 30 GB of RAM, as it seemed to me that HDBSCAN / UMAP on GPU with CUML is faster than the CPU. With the CPU I could get RAM until like 512 GB, but 1,4 TB still seems way to high? Especially in relation to me only working on a dataset of 440k small images…

Do you have any idea on that?

```
2023-08-10 01:24:34,343 - BERTopic - Reduced dimensionality
2023-08-10 01:26:43,227 - BERTopic - Clustered reduced embeddings
Traceback (most recent call last):
  File "/efs/users/groesch/ondemand/buildFModelBasedOnEmbeddingsQuer.py", line 35, in <module>
    topics = topic_model.fit_transform(documents=None, embeddings=embeddingsLoaded, images=embeddings_joint_imageslist)
  File "/users/groesch/.local/lib/python3.10/site-packages/bertopic/_bertopic.py", line 397, in fit_transform
    custom_documents = self._images_to_text(documents, embeddings)    
  File "/users/groesch/.local/lib/python3.10/site-packages/bertopic/_bertopic.py", line 3140, in _images_to_text
    documents = tuner.image_to_text(documents, embeddings)
  File "/users/groesch/.local/lib/python3.10/site-packages/bertopic/representation/_visual.py", line 167, in image_to_text
    indices = mmr(topic_embedding.reshape(1, -1), embeddings[indices], indices, top_n=top_n, diversity=0.1)
  File "/users/groesch/.local/lib/python3.10/site-packages/bertopic/representation/_mmr.py", line 96, in mmr
    word_similarity = cosine_similarity(word_embeddings)
  File "/opt/apps/hand/anaconda3/202303/lib/python3.10/site-packages/sklearn/metrics/pairwise.py", line 1401, in cosine_similarity
    K = safe_sparse_dot(X_normalized, Y_normalized.T, dense_output=dense_output)
  File "/opt/apps/hand/anaconda3/202303/lib/python3.10/site-packages/sklearn/utils/extmath.py", line 189, in safe_sparse_dot
    ret = a @ b
numpy.core._exceptions._ArrayMemoryError: Unable to allocate 1.40 TiB for an array with shape (438365, 438365) and data type float64
```

**Comment by MaartenGr** [comment]
Could you share your full code? That makes it easier to understand what exactly is happening. Also, how big is your training data?

**Comment by gr0esch** [comment]
Hey @MaartenGr, thanks for the quick reply!

Certainly, here's the full code.

With the error message above the min_samples was set to 10, when changing to 100 it is "only" asking for 405 GB of RAM. When I'll run this on the CPU it will be way slower, right?

```
from cuml.cluster import HDBSCAN
from cuml.manifold import UMAP

from bertopic.representation import KeyBERTInspired, VisualRepresentation
from bertopic.backend import MultiModalBackend
from bertopic import BERTopic
import numpy as np

import pickle


hdbscan_model = HDBSCAN(min_samples=100, gen_min_span_tree=True, prediction_data=False)
umap_model = UMAP(n_components=5, n_neighbors=15, min_dist=0.0, low_memory=True)


# Image to text representation model
representation_model = {
    "Visual_Aspect": VisualRepresentation(image_to_text_model="nlpconnect/vit-gpt2-image-captioning")
}

# Image embedding model
embedding_model = MultiModalBackend('clip-ViT-B-32', batch_size=32)

embeddingsLoaded = pickle.load(open("embeddings_quer.pkl", 'rb')) #To load saved model from local directory

embeddings_joint_imageslist = pickle.load(open("embeddings_quer_imageslist.pkl", 'rb')) #To load saved model from local directory

# Train our model with images only
topic_model = BERTopic(low_memory=True,verbose=True,calculate_probabilities=False,embedding_model=embedding_model,representation_model=representation_model,umap_model=umap_model,hdbscan_model=hdbscan_model)


topics = topic_model.fit(documents=None, embeddings=embeddingsLoaded, images=embeddings_joint_imageslist)

pickle.dump(topics, open("embeddings_quer_fullmodel.pkl", 'wb')) #Saving images
```

**Comment by MaartenGr** [comment]
I believe what is happening is the following. When you have created clusters of images a subset of images is selected in each cluster to convert to text, to caption. To do this, Maximal Marginal Relevance is used to select the top 9 images that are close to the centroid but are still somewhat diverse. However, it calculates the cosine similarity between the images themselves to do so, resulting in a matrix of 438365 by 438365. This value of 438365 refers to the number of documents in your cluster. 

If you change `min_samples` that might decrease the size of the largest cluster. But with your large dataset, you might still create large clusters resulting in these large similarity matrices. 

To prevent them, there would need to be the fix needed I believe. For example, instead of using Maximal Marginal Relevance on all images within a cluster, it might suffice to do this for only 10000 or so, thereby creating matrices at most 10000 by 10000. It would, however, make the resulting image representations a tad less accurate by I think with a subset of 10000 this effect would hardly be noticeable. 

In other words, could you change the following line:

https://github.com/MaartenGr/BERTopic/blob/37064e231defceb998ef2a778cef46fc769dbf00/bertopic/representation/_visual.py#L167

To this:

```python
indices = mmr(topic_embedding.reshape(1, -1), embeddings[indices[:10_000]], indices[:10_000], top_n=top_n, diversity=0.1)
```

And try again?

**Comment by gr0esch** [comment]
Thanks, this sound not only very reasonable, but the quick code adoption worked.

Do you mind leaving that ticket open for a few more days, I might run into an issue with the bigger dataset.

**Comment by gr0esch** [comment]
And that is exactly happening… So I ran my code yesterday on the medium and small one, everything worked fine. Started the job for the large one (pkl file with embeddings nearly 10 GB or 2,5 million image files) but nothing is happening :(

The output starts the same way as with the other two sets, though it is not even printing the dimension reduction has finished. So I get the warning message, that I had with the other ones as well (and indicated no problem then), but after that, no further output, while the job is now running for 15,5 hours. 

The only difference between this set and the two other ones is the size and that while calculating the embeddings, for the large set I split the images in large batches as otherwise I ran into OOM, which is why I have to concatenate the loaded pkl first. Though, no errors raised there.

Am I too impatient? I'm running this now on a CPU with 240 GB RAM with no OOM indicated. Tried it before on smaller units and got OOM straight away, so that should be fine now?

Do you have any ideas / suggestions? Thank you very much!

Full output
```
/users/groesch/.local/lib/python3.10/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.[0m
  @numba.jit()
/users/groesch/.local/lib/python3.10/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.[0m
  @numba.jit()
/users/groesch/.local/lib/python3.10/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.[0m
  @numba.jit()
/users/groesch/.local/lib/python3.10/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: [1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.[0m
  @numba.jit()
ftfy or spacy is not installed using BERT BasicTokenizer instead of ftfy.
```
Full code

```pythone
from hdbscan import HDBSCAN
from umap import UMAP

from bertopic.representation import KeyBERTInspired, VisualRepresentation
from bertopic.backend import MultiModalBackend
from bertopic import BERTopic
import numpy as np

import pickle


hdbscan_model = HDBSCAN(min_samples=10, gen_min_span_tree=True, prediction_data=True)
umap_model = UMAP(n_components=5, n_neighbors=15, min_dist=0.0, low_memory=True)


# Image to text representation model
representation_model = {
    "Visual_Aspect": VisualRepresentation(image_to_text_model="nlpconnect/vit-gpt2-image-captioning")
}

# Image embedding model
embedding_model = MultiModalBackend('clip-ViT-B-32', batch_size=32)

embeddingsLoaded = pickle.load(open("embeddings_joint.pkl", 'rb')) #To load saved model from local directory

embeddings_joint_imageslist = pickle.load(open("embeddings_joint_imageslist.pkl", 'rb')) #To load saved model from local directory

# Join all the embedding batches into a single list
#flattened_embeddings = [embedding for batch_embeddings in embeddingsLoaded for embedding in batch_embeddings]

embeddingsConc = np.concatenate(embeddingsLoaded, axis=0)


# Train our model with images only
topic_model = BERTopic(low_memory=True,verbose=True,calculate_probabilities=False,embedding_model=embedding_model,representation_model=representation_model,umap_model=umap_model,hdbscan_model=hdbscan_model)


topics = topic_model.fit_transform(documents=None, embeddings=embeddingsConc, images=embeddings_joint_imageslist)

pickle.dump(topics, open("embeddings_joint_fullmodel.pkl", 'wb')) #Saving images

embedding_model_name = "clip-ViT-B-32"
topics.save("embeddings_joint_fullmodel_safetensors", serialization="safetensors", save_ctfidf=True, save_embedding_model=embedding_model_name)
topics.save("embeddings_joint_fullmodel_pytorch", serialization="pytorch", save_ctfidf=True, save_embedding_model=embedding_model_name)
```

**Comment by MaartenGr** [comment]
You are by no means impatient! But it does seem that you switched over from cuML to the original UMAP implementation. If you want to handle large datasets for dimensionality reduction, then cuML together with a strong GPU is a necessity. Also, it is important to note that not everything scales linearly, including UMAP, scales linearly with size so it would not be unsurprising if the computation time also does not scale linearly. 

All in all, I would advise going with cuML instead. Moreover, make sure that you set `min_cluster_size` together with `min_samples` in HDBSCAN relatively large (a couple of hundred). Otherwise, there is a very good chance you will end up with tens of thousands of topics. Lastly, check whether a `cosine` metric is possible within cuML's UMAP. If not, then normalization of embeddings might be necessary.

**Comment by gr0esch** [comment]
Yes, I switched to the original UMAP & HDBSCAN because I was running that one on CPU, without GPU available cuML would fail…? 

I think min_samples was limited to something above 100, got an error like that when I tried to set it like 1000, I remember.
I'll now give it a go with 128 GB of RAM on a GPU with cuML, `min_samples=100` and, `min_cluster_size=500`

According to the docs it seems that cosine is not possible with cuML, I therefore integrate your snippet. Do you mind elaborating a little bit why this makes sense? Would help me a lot! I'll let you know if this setup works.

```python
from cuml.preprocessing import normalize
embeddings = normalize(embeddings)
```

**Comment by gr0esch** [comment]
I now have a really really weird behaviour, that I can't explain to myself…

Gave it a shot with the big instance but that seemed to got stuck even before actually reducing the dimensions. So I tested how long it would take to actually load the pkl, found out – only seconds. Changed the code to have a progress bar in the pkl import and added some printout…

Soo it works very fine as long as quit() before importing the BERTopic part, the log is printed an so on. But when removing the quit(), the script continues to run without any error message but after `
embeddings_joint_imageslist = test_tqdm_reader("embeddings_joint_imageslist.pkl")` it gets stuck. `"Loaded Embeddings"`and the progressbar of the `embeddings_joint_imageslist.pkl` is printed, but after that – nothing, without any error the job keeps running but is not doing anything?!

Any ideas?

```python


print("STARTING")
import numpy as np
import os

from cuml.preprocessing import normalize


class TQDMBytesReader(object):

    def __init__(self, fd, **kwargs):
        self.fd = fd
        from tqdm import tqdm
        self.tqdm = tqdm(**kwargs)

    def read(self, size=-1):
        bytes = self.fd.read(size)
        self.tqdm.update(len(bytes))
        return bytes

    def readline(self):
        bytes = self.fd.readline()
        self.tqdm.update(len(bytes))
        return bytes

    def __enter__(self):
        self.tqdm.__enter__()
        return self

    def __exit__(self, *args, **kwargs):
        return self.tqdm.__exit__(*args, **kwargs)

       
def test_tqdm_reader(file):
    from pickle import Unpickler
    with open(file, "rb") as fd:
        total = os.path.getsize(file)
        with TQDMBytesReader(fd, total=total) as pbfd:
            up = Unpickler(pbfd)
            obj = up.load()
            return(obj)
        #print(f"Loaded {str(obj)}")
        
print("Will now load embeddings")

embeddingsLoaded = test_tqdm_reader("embeddings_joint.pkl")

print("Loaded Embeddings")

embeddings_joint_imageslist = test_tqdm_reader("embeddings_joint_imageslist.pkl")  

print("Loaded Images List")

embeddingsConc = np.concatenate(embeddingsLoaded, axis=0)
print("Concatenated")
print(len(embeddingsConc))
embeddingsConc = normalize(embeddingsConc)
print("Normalized")

quit()

from cuml.cluster import HDBSCAN
from cuml.manifold import UMAP

from bertopic.representation import KeyBERTInspired, VisualRepresentation
from bertopic.backend import MultiModalBackend
from bertopic import BERTopic

hdbscan_model = HDBSCAN(min_samples=100,min_cluster_size=500,gen_min_span_tree=True, prediction_data=True)
print("Created HDBSCAN")
umap_model = UMAP(n_components=5, n_neighbors=15, min_dist=0.0, low_memory=True)
print("Created UMAP")

# Image to text representation model
representation_model = {
    "Visual_Aspect": VisualRepresentation(image_to_text_model="nlpconnect/vit-gpt2-image-captioning")
}

print("Created RepresntationModel")

# Image embedding model
embedding_model = MultiModalBackend('clip-ViT-B-32', batch_size=32)
print("Created CLIP Model")

# Join all the embedding batches into a single list
#flattened_embeddings = [embedding for batch_embeddings in embeddingsLoaded for embedding in batch_embeddings]


# Train our model with images only
topic_model = BERTopic(low_memory=True,verbose=True,calculate_probabilities=False,embedding_model=embedding_model,representation_model=representation_model,umap_model=umap_model,hdbscan_model=hdbscan_model)
print("Full BERT Model initated")

topics = topic_model.fit(documents=None, embeddings=embeddingsConc, images=embeddings_joint_imageslist)

print("Model fitted")

embedding_model_name = "clip-ViT-B-32"
topics.save("embeddings_joint_fullmodel_safetensors", serialization="safetensors", save_ctfidf=True, save_embedding_model=embedding_model_name)
print("Saved tensors")
topics.save("embeddings_joint_fullmodel_pytorch", serialization="pytorch", save_ctfidf=True, save_embedding_model=embedding_model_name)
print("Saved pytorch")
pickle.dump(topics, open("embeddings_joint_fullmodel.pkl", 'wb')) #Saving images
print("Saved pkl")

```

**Comment by MaartenGr** [comment]
> Yes, I switched to the original UMAP & HDBSCAN because I was running that one on CPU, without GPU available cuML would fail…?

The original UMAP cannot easily handle millions of datapoints, so cuML's UMAP would be preferred if you have a good GPU available. Generally, a GPU is highly advised since you are using a bunch of different LLMs throughout text/image analyses. 

> According to the docs it seems that cosine is not possible with cuML, I therefore integrate your snippet. Do you mind elaborating a little bit why this makes sense? Would help me a lot! I'll let you know if this setup works.

You can find a nice explanation for this [here](https://stats.stackexchange.com/a/299016/224272). 

> Any ideas? What I can not explain is the fact, that this seems to be influenced by code that is only running after that. With the quit(), everything seems to process until then, without, it even stops before.

I am not sure what is happening there. But then again, it does not seem to be BERTopic-related since you created a custom tqdm-related reader. So issues with that would be best handled in the respective repo.
